core/drivers/stm32_cpu_opp.c

```diff
@@ -0,0 +1,354 @@
+static struct mutex cpu_opp_mu = MUTEX_INITIALIZER;
+
+size_t stm32_cpu_opp_count(void)
+{
+	return cpu_opp.opp_count;
+}
+
+/* Perf level relates straight to CPU frequency in kHz */
+unsigned int stm32_cpu_opp_level(unsigned int opp)
+{
+	assert(opp < cpu_opp.opp_count);
+
+	return cpu_opp.dvfs[opp].freq_khz;
+}
+
+static TEE_Result set_opp_clk_rate(unsigned int opp)
+{
+	return clk_set_rate(cpu_opp.clock, cpu_opp.dvfs[opp].freq_khz * 1000);
+}
+
+static TEE_Result set_opp_voltage(unsigned int opp)
+{
+	return regulator_set_voltage(cpu_opp.regul, cpu_opp.dvfs[opp].volt_uv);
+}
+
+static TEE_Result set_clock_then_voltage(unsigned int opp)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = set_opp_clk_rate(opp);
+	if (res)
+		return res;
+
+#ifdef CFG_STM32MP13
+	if (cpu_opp.dvfs[opp].volt_uv <= PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+		io_setbits32(stm32_pwr_base() + PWR_CR1_OFF,
+			     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+	res = set_opp_voltage(opp);
+	if (res) {
+		/* Restore previous OPP */
+#ifdef CFG_STM32MP13
+		if (cpu_opp.dvfs[cpu_opp.current_opp].volt_uv >
+		    PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+			io_clrbits32(stm32_pwr_base() + PWR_CR1_OFF,
+				     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+		if (set_opp_clk_rate(cpu_opp.current_opp))
+			panic();
+
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result set_voltage_then_clock(unsigned int opp)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = set_opp_voltage(opp);
+	if (res)
+		return res;
+
+#ifdef CFG_STM32MP13
+	if (cpu_opp.dvfs[opp].volt_uv > PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+		io_clrbits32(stm32_pwr_base() + PWR_CR1_OFF,
+			     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+	res = set_opp_clk_rate(opp);
+	if (res) {
+		/* Restore previous OPP */
+#ifdef CFG_STM32MP13
+		if (cpu_opp.dvfs[cpu_opp.current_opp].volt_uv <=
+		    PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+			io_setbits32(stm32_pwr_base() + PWR_CR1_OFF,
+				     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+		if (set_opp_voltage(cpu_opp.current_opp))
+			panic();
+
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result stm32_cpu_opp_set_level(unsigned int level)
+{
+	unsigned int current_level = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int opp = 0;
+
+	DMSG("Set OPP to %ukHz", level);
+
+	mutex_lock(&cpu_opp_mu);
+
+	current_level = stm32_cpu_opp_level(cpu_opp.current_opp);
+
+	if (level == current_level) {
+		mutex_unlock(&cpu_opp_mu);
+		return TEE_SUCCESS;
+	}
+
+	for (opp = 0; opp < cpu_opp.opp_count; opp++)
+		if (level == cpu_opp.dvfs[opp].freq_khz)
+			break;
+
+	if (opp == cpu_opp.opp_count) {
+		mutex_unlock(&cpu_opp_mu);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (level < current_level)
+		res = set_clock_then_voltage(opp);
+	else
+		res = set_voltage_then_clock(opp);
+
+	if (res)
+		EMSG("Failed to set OPP to %ukHz", level);
+	else
+		cpu_opp.current_opp = opp;
+
+	mutex_unlock(&cpu_opp_mu);
+
+	return res;
+}
+
+TEE_Result stm32_cpu_opp_read_level(unsigned int *level)
+{
```

Comment:
For consistency, here also we should check the OPPs are defined:
```c
if (!cpu_opp.opp_count) {
    EMSG("No CPU OPP defined");
    return TEE_ERROR_GENERIC;
}

Reply:
My bad: I may have been confusing. Testing whether OPP support was initialized or not, we should test a unique pattern. This function check `cpu_opp.opp_count != 0` whereas `stm32_cpu_opp_set_level()` checks `cpu_opp.dvfs != NULL`. On the ohter hand, `stm32_cpu_opp_get_dt_subnode()` only resets `cpu_opp.opp_count`, not `cpu_opp_dvfs` upon failure. Could you make all these more consistent?